tests/selfdestruct_test.go

```diff
@@ -0,0 +1,394 @@
+package tests
+
+import (
+	"math/big"
+	"testing"
+
+	"github.com/Fantom-foundation/go-opera/tests/contracts/query_account"
+	"github.com/Fantom-foundation/go-opera/tests/contracts/selfdestruct"
+	"github.com/Fantom-foundation/go-opera/tests/contracts/selfdestruct_factory"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSelfDestruct(t *testing.T) {
+	require := require.New(t)
+
+	someBalance := big.NewInt(1234)
+	addressCount := 1234 // to avoid beneficiary address collisions
+
+	net, err := StartIntegrationTestNet(t.TempDir())
+	require.NoError(err, "failed to start test network")
+	defer net.Stop()
+
+	// Utility contract to query a second account state
+	queryAccount, receipt, err := DeployContract(net, query_account.DeployQueryAccount)
+	require.NoError(err)
+	require.Equal(receipt.Status, types.ReceiptStatusSuccessful, "failed to deploy contract")
+
+	t.Run("constructor", func(t *testing.T) {
+		tests := map[string]struct {
+			deployTx  deployTxFunction[selfdestruct.Selfdestruct]
+			executeTx executeTxFunction[selfdestruct.Selfdestruct]
+			effects   map[string]effectFunction
+		}{
+			// This test checks the testing infrastructure itself
+			"sanity check/no selfdestruct": {
+				deployTx: func(opts *bind.TransactOpts, backend bind.ContractBackend, otherAddress common.Address) (common.Address, *types.Transaction, *selfdestruct.Selfdestruct, error) {
+					opts.Value = someBalance
+					return selfdestruct.DeploySelfdestruct(opts, backend,
+						false, // do not selfdestruct in constructor
+						false, // beneficiary is not self
+						otherAddress)
+				},
+				effects: map[string]effectFunction{
+					"contract keeps balance":       contractBalanceIs(someBalance),
+					"beneficiary gains no balance": beneficiaryBalanceIs(big.NewInt(0)),
+					"storage is not deleted":       contractStorageIs(big.NewInt(123)),
+					"code is not deleted":          contractCodeSizeIsNot(big.NewInt(0)),
```

Comment:
I tried to have some placeholder variable, to be modified afterward by the test execution, filled in with the assigned contract address. After a few iterations, I gave up. The pointers and copy of the value by the lambdas is difficult to debug, the current implementation constructs values to be used once by the matchers and has less moving parts that could go wrong.

This infrastructure was built because the first unit test I did, did not scale and produced many problems.
The big wins of this approach are: 
- contracts are used once per test. To prevent state-keeping in the blockchain
- beneficiary is a new address for each test, to avoid having to reset its balance (which would require a contract) 
- declarative definition of the test cases, to reduce boilerplate between effects.